### PR TITLE
Upfront Wizard

### PR DESCRIPTION
**Pull request by m50**
Implements #24.

This adds an Operating System preference to the User, and allows you to set a track and operating system and locale.

For saving your locale, we need to wait for #103. 

Todo:
* [x] Filter display based on your preferred Operating System.
* [x] Save preferred locale.
* [x] Add a user preferences page to be able to change these at a later date.

### CODE REVIEW

**Comment by mattstauffer** [comment]
@m50 Sorry I'm holding you up. #103 is ... a beast. :) Thanks for your work here--I can't wait to review it!

File: app/Http/Controllers/Auth/WizardController.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    public function __invoke()
+    {
+        $valid = $this->validate(request(), [
+            'os' => ['required', 'string', 'in:any,osx,linux,windows'],
+            'track' => ['required', 'int', 'exists:tracks,id', 'min:1'],
+            'locale' => ['required', 'string', Rule::in((new Locale())->slugs())]
+        ], [], [
+            'os' => 'OS'
+        ]);
+        $user = auth()->user();
```
Comment (mattstauffer) [inline app/Http/Controllers/Auth/WizardController.php]: We might be able to simplify this:

```php
auth()->user()->update([
    'os' => $valid['os'],
    'track' => $valid['track'],
]);
```

File: app/Http/Controllers/Auth/WizardController.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        $user->os = $valid['os'];
+        $user->track_id = $valid['track'];
+        // TODO: Add language save. Waiting on PR#103.
+        $user->save();
+        return redirect(path_wlocale($this->redirectTo));
```
Comment (mattstauffer) [inline app/Http/Controllers/Auth/WizardController.php]: Newline above return please

File: database/factories/ResourceFactory.php [hunk 3]
```diff
@@ -13,5 +13,6 @@
         'type' => $faker->randomElement(['video', 'course', 'audio', 'book', 'article']),
         'module_id' => factory(Module::class),
         'language' => $faker->randomElement(['en', 'es']),
+        'os' => $faker->randomElement(['any', 'osx', 'linux', 'windows']),
```
Comment (mattstauffer) [inline database/factories/ResourceFactory.php]: I think we'll probably want to make an enum for this. 

I haven't yet 100% decided how I want to build the enums but here's what I've been thinking:

```php
<?php

namespace App;

use Illuminate\Contracts\Support\Arrayable;

abstract class Enum implements Arrayable
{
    abstract public function keys();

    public function values()
    {
        return collect($this->keys())->map(function ($key) {
            return static::$key;
        })->toArray();
    }

    public function toArray()
    {
        return collect($this->keys())->mapWithKeys(function ($key) {
            return [$key => static::$key];
        })->toArray();
    }
}
```
And:

```php
<?php
namespace App;

class OperatingSystem extends Enum
{
    const MACOS = 'macos';
    const LINUX = 'linux';
    const WINDOWS = 'windows';

    public static function keys()
    {
        return [
            'MACOS',
            'LINUX',
            'WINDOWS',
        ];
    }
}

File: app/Http/Controllers/Auth/WizardController.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+class WizardController extends Controller
+{
+    protected $redirectTo = 'home';
+    
+    public function __invoke()
+    {
+        $valid = $this->validate(request(), [
+            'os' => ['required', 'string', 'in:any,osx,linux,windows'],
+            'track' => ['required', 'int', 'exists:tracks,id', 'min:1'],
```
[the remark this answers is not in the document]
Comment (m50) [reply app/Http/Controllers/Auth/WizardController.php]: No, it was there before I put the exists in and hadn't cleaned it up.

**Comment by m50** [comment]
Well, the rest of this is relying on #103 to be able to properly implement.

Should I create the preferences page, or is that already being worked on?

**Comment by m50** [comment]
Sounds good, I'll merge #103 into this branch and continue so.

**Comment by m50** [comment]
I did one extra thing. If already logged in, going to '/' will redirect to your prefered locale. Also, after logging in, it'll redirect to your preferred locale, rather than whatever the current is, in case you log in on the english page, but your preferred locale is swedish or whatever.

**Comment by mattstauffer** [comment]
Thanks for your work on this @m50!

**Pull request closed without merging**